pkg/segment/set.go

```diff
@@ -932,11 +1001,203 @@ func (n *node) rebalanceAfterRemove(gap GapIterator) GapIterator {
+			if parentNewMax != n.parent.maxGap {
+				n.parent.updateMaxGapInternal(parentNewMax)
+			}
+		} else {
+			// If new max is smaller than the old maxGap, but parent's maxGap is not this one,
+			// only need to update locally
+			n.maxGap = max
+		}
+	}
+}
+
+// updateMaxGapInternal updates the maxGap on internal nodes recursively.
+//
+// Precondition: Should only be called by updateMaxGapLeaf or updateMaxGapInternal recursively.
+func (n *node) updateMaxGapInternal(newMaxGap Key) {
+	if !n.hasChildren {
+		panic(fmt.Sprintf("updateMaxGapInternal should always be called on non-leaf node: %v", n))
+	}
+	if newMaxGap == n.maxGap {
+		// If newMaxGap equals the old maxGap, no update is needed.
+		return
+	} else if newMaxGap > n.maxGap {
+		// If newMaxGap is larger than the old maxGap, update it.
+		n.maxGap = newMaxGap
+		// Only if new maxGap is larger than parent's old maxGap, propagate this update to parent.
+		if n.parent != nil && n.parent.maxGap < n.maxGap {
+			n.parent.updateMaxGapInternal(newMaxGap)
+		}
+	} else {
+		if n.parent != nil && n.parent.maxGap == n.maxGap {
+			// If newMaxGap is smaller than the old maxGap, and this gap used to be the maxGap of its parent,
+			// iterate parent's children and calculate parent's new maxGap.
+			// (It's probable that parent has two children with the old maxGap, but we need to check it anyway.)
+			n.maxGap = newMaxGap
+			var parentNewMax Key
+			for i := 0; i <= n.parent.nrSegments; i++ {
+				child := n.parent.children[i]
+				if temp := child.maxGap; i == 0 || temp > parentNewMax {
+					parentNewMax = temp
+				}
+			}
+			// If parent's new maxGap differs from the old one, propagate this update.
+			if parentNewMax != n.parent.maxGap {
+				n.parent.updateMaxGapInternal(parentNewMax)
+			}
+		} else {
+			// If newMaxGap is smaller than the old maxGap, but parent's maxGap is not this one,
+			// only need to update locally
+			n.maxGap = newMaxGap
+		}
+	}
+}
+
+// updateLocalMaxGap updates maxGap of the calling node solely with no propagation to ancestor nodes.
+func (n *node) updateLocalMaxGap() {
+	var max Key
+	if !n.hasChildren {
+		// leaf node iterates gaps
+		for i := 0; i <= n.nrSegments; i++ {
+			currentGap := GapIterator{n, i}
+			if temp := currentGap.Range().Length(); i == 0 || temp > max {
+				max = temp
+			}
+		}
+	} else {
+		// non-leaf node iterates its children
+		for i := 0; i <= n.nrSegments; i++ {
+			child := n.children[i]
+			if temp := child.maxGap; i == 0 || temp > max {
+				max = temp
+			}
+		}
+	}
+	n.maxGap = max
+}
+
+// searchFirstLargeEnoughGap returns the first gap having at least minSize length
+// in the subtree rooted by n. If not found, return a terminal gap iterator.
+func (n *node) searchFirstLargeEnoughGap(minSize Key) GapIterator {
+	if n.maxGap < minSize {
+		return GapIterator{}
+	}
+	if n.hasChildren {
+		for i := 0; i <= n.nrSegments; i++ {
+			if temp := n.children[i].searchFirstLargeEnoughGap(minSize); temp.Ok() {
```

Comment:
Here and elsewhere: use a descriptive variable name and not "temp". Is it the gap iterator? Call it "gi" or just "gap".